tests/cpp/test_cache.cc

```diff
@@ -0,0 +1,52 @@
+/**
+ * Copyright 2023 by XGBoost contributors
+ */
+#include <gtest/gtest.h>
+#include <xgboost/cache.h>
+#include <xgboost/data.h>  // DMatrix
+
+#include <cstddef>         // std::size_t
+
+#include "helpers.h"       // RandomDataGenerator
+
+namespace xgboost {
+namespace {
+struct CacheForTest {
+  std::size_t i;
+
+  explicit CacheForTest(std::size_t k) : i{k} {}
+};
+}  // namespace
+
+TEST(DMatrixCache, Basic) {
+  std::size_t constexpr kRows = 2, kCols = 1, kCacheSize = 4;
+  DMatrixCache<CacheForTest> cache(kCacheSize);
+
+  auto add_cache = [&]() {
+    auto p_fmat = RandomDataGenerator(kRows, kCols, 0).GenerateDMatrix();
```

Comment:
```suggestion
    // Create a lambda function here, so that p_fmat gets deleted upon the
    // end of the lambda. This is to test how the cache handle expired
    // cache entries.
    auto p_fmat = RandomDataGenerator(kRows, kCols, 0).GenerateDMatrix();
```